### ISSUE TITLE
Provide the tools to pass devices (and the rest) around across Send/Sync boundaries

### ISSUE THREAD

**Issue by MOZGIII** [RustAudio/cpal#435]
`Device`s and `Stream`s aren't `Send`/`Sync`.

To pass them around, we can create a thread to bind them all to, and expose channel-based control interface over it to allow passing commands to streams and devices.

This does *not* have to be baked into the device implementation, but can be a separate utility for users to manually use if they need to.

The reason to include it into `cpal` is it's that non `Send`/`Sync` is a very serious limitation, and it should be solved at the library level, since the underlying design limitation also comes from the library level.

**Comment by sniperrifle2004** [comment]
I don't know. It seems like a bad idea (`Send` maybe, but `Sync`?). Stream might be `Sync` (Might be some implementation caveats, but it's likely doable) but `Device` is just asking for trouble. Device is already not completely safe (I can start multiple streams at the same time. It fails miserably, but `Device` is not preventing me from doing so) and you want to add `Sync` to that?

**Comment by MOZGIII** [comment]
What I have in mind is a simple thread holding to a non-`Send`/`Sync` `Device`(s) with a state machine and a set of channels to pass commands in and extract events out. It's what I can already implement currently with existing interfaces, but I figured it would be something useful to have at the lib level.

I wasn't aware it's problematic to have multiple streams in one direction with a single `Device`. Clearly, it's not by design, but I wonder how it's broken to avoid relying on that - and maybe to land a workaround in place. Could you please link an issue number on this bug for context?

**Comment by sniperrifle2004** [comment]
Well it was the first problematic scenario that popped into my head with a `Sync` device (The full featured `Device`) so I tried it out in the current situation. Hence there isn't an issue for it yet (As far as I'm aware).

I think the creation command is subject to the same problems as the current method is. So it would still require answers to those questions. Don't those stream handles become stale?

**Comment by sniperrifle2004** [comment]
Ah I think I understand what is going on. Your device is capable of mixing and that's why for you it seems perfectly normal for more streams to be running on the same device. Now I was addressing my card directly and thus it failed, since my sound card doesn't have a mixer. Now there are more potential problems I can think of with concurrent streams based on the interface (The device would also need to be able to handle different configs at the same time). So for me it wasn't at all obvious that multiple streams were supported by design.

**Comment by MOZGIII** [comment]
> Ah I think I understand what is going on. Your device is capable of mixing and that's why for you it seems perfectly normal for more streams to be running on the same device. Now I was addressing my card directly and thus it failed, since my sound card doesn't have a mixer. Now there are more potential problems I can think of with concurrent streams based on the device interface (The device would also need to be able to handle different configs at the same time). So for me it wasn't at all obvious that multiple streams were supported by design. To be honest it still isn't. At least not inherently supported.

I'm not targeting a specific sound card, but rather building a portable cross-platform app. It's generally supported across all platforms to have multiple streams (even on Linux PulseAudio emulates ALSA interface and does the mixing for the system).

But then, I don't care at all about having multiple streams in two directions, I'll only be using one capture and one playback stream. My problem is purely that I want to be able to synchronize access to `cpal` from multiple threads - ideally in an `async` fashion.

I do some mixing internally, but that's totally unrelated to the `cpal` layer. Also, the callback interfaces are already fine - I'm having no issues cause I have `Send` ringbuffers with disjoint writer and reader ends. My problem is just that when I do the device initialization and format negotiation - I already need to cross `Send` boundaries a couple of times before I can spawn a stream. Thus, I don't care as much about the `Sync` either.

It used to work fine with an event loop, because then the `Device` didn't have those limitations.

**Comment by sniperrifle2004** [comment]
> 
> I'm not targeting a specific sound card, but rather building a portable cross-platform app. It's generally supported across all platforms to have multiple streams (even on Linux PulseAudio emulates ALSA interface and does the mixing for the system).
> 
But I was, which is why it failed. The "default" card is often some software over hardware that does the mixing. When I changed it back to my "default" device it worked. Now I had never heard of the mixing of said device being relied on by an application rather than just the mixing between different applications before. I'm not saying `Device` can't possibly support multiple output streams. Just that `Device` doesn't in and of itself provided. In order to do that the underlying device must do the mixing.

**Comment by MOZGIII** [comment]
This is going slightly off-topic. I suggest you create another issue because having a wrapper `Device` that could do mixing and resampling would be useful too. I need a simpler thing though.

**Comment by sniperrifle2004** [comment]
I don't think you do. Once you hit a `Device` that doesn't do mixing your simpler thing will not solve your problem any more and your application will not work. Granted that might be unlikely, but in the broad cpal sense that unlikely is a problem.

**Comment by MOZGIII** [comment]
I don't want `cpal` to do mixing for my current use case. I *do* have mixing/resampling/transcoding facilities, but I want them to be separate from `cpal` and entirely under my control, rather than the library control.

I need a convenient way to cross `Send` borders. That's it.

**Comment by roderickvd** [comment]
The current v0.17 and soon-to-be-released v0.18 already had it for most hosts. #1221 implements it for all hosts on the `develop` branch slated for a future v0.19.
